Feature/tlt 2522 selenium bulk create registrar code

hktest opened the pull request:
Tested locally and successfully.  
These adds additional tests to check that the course code field is never null in bulk create tool.  Every course should have course code or registrar_code_display.
 
The desired tool display logic for TLT-2522 is that 
- if there is no registrar_code_display value, the  course code is displayed in the bulk create tool's  course code column, but 
- if registrar_code_display has a value, the registrar_code_display value should show up in the bulk create tool's course code column.
Set up test data to mirror these two scenario.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -35,4 +50,57 @@ def test_course_selection(self):
+
+    def test_course_code_is_not_null_if_there_is_a_course_code(self):
+        """
+        Tests that a course code field is not null, if there is a course code
```
I'm pretty confused by the descriptions of these test cases. I think it would be helpful to sit down and make sure we're testing what we're supposed to be testing. The `course_code_no_registrar_display` and `registrar_code_display_text` seem to have course instance IDs or something in them, instead of course registrar codes, and we might be testing the wrong things...?

hktest replied:
Getting rid of a few of the methods, per feedback.  Will look into better description of both test data and test methods to see if that helps to resolve some of the confusion.

hktest replied:
fixed -- added comments and renamed local variables and methods to have a clearer description.  hopefully less confusing after these changes?

elliottyates commented on icommons_lti_tools/settings/local.py:
```diff
@@ -16,6 +16,13 @@
             'template': 'None',  # No Template
             'term': '1',
+
+            'course_code_no_registrar_display': '109581',
```
Can we keep this sorted alphabetically?

elliottyates commented on selenium_tests/bulk_create/page_objects/index_page.py:
```diff
@@ -16,11 +16,20 @@ class Locators(object):
+        """
+        returns a locator for course code display text in the table
+        """
+        # print "//td[contains(text(),'%s')]" % course_code_text
```
looks like we can remove this now

hktest replied:
Removed.

elliottyates commented on selenium_tests/bulk_create/page_objects/index_page.py:
```diff
@@ -16,11 +16,20 @@ class Locators(object):
+        returns a locator for course code display text in the table
+        """
+        # print "//td[contains(text(),'%s')]" % course_code_text
+        return By.XPATH, "//td[contains(text(),'%s')]" % course_code_text
```
Probably better to do an exact match, instead of `contains`.

Joy603 replied:
You can probably use '"{}" .format() syntax, instead of %s.

hktest replied:
done, thanks josie.  used your suggested, "return By.XPATH, "//td[text()='{}']".format(course_code_text)"  format.

elliottyates commented on selenium_tests/bulk_create/page_objects/index_page.py:
```diff
@@ -55,3 +64,49 @@ def get_new_course_link(self):
         element = self.find_element(*Locators.CREATE_NEW_COURSE_LINK)
         element.click()
+
+    def check_course_code_is_not_null_course_code_exists(self,
```
These four methods don't really map to general services that the page offers -- they are instead specific assertions of test cases, from what I can see. POs provide services that are independent of the thing you're checking on the page -- they can be reused in different tests for different reasons. For instance, a single test that checks the table for a TD with some text would accomplish what you need here -- in that case, `COURSE_CODE_TEXT` becomes `TD_WITH_TEXT` (since it's just searching for a td) and you have a single service (method) called, e.g. `table_contains()`, which you then use in the test case to assert these four cases. We don't need four separate services, as the PO just needs to provide a single service that meets all four use cases.

hktest replied:
Very good point.  I did this one all wrong (didn't feel quite right either with the 4 similar methods, but I coudn't quite put my finger on how to fix it -- in code).   Referenced Josie's PR to get some ideas on how that got done - made sense.  Fixed!

elliottyates commented:
I think a lot of my confusion comes from nomenclature. What the story calls the 'course code' is actually in our database called the CI's course.registrar_code vs. the registrar_code_display, which should be shown if it exists (else fall back on the registrar_code).

In [this test code](https://github.com/Harvard-University-iCommons/icommons_lti_tools/pull/447/files#diff-8c6d76a66ffb8d75de8eefbf51aca455R15), though:
```
'course_code_text': '109581',
'course_code_no_registrar_display': '109581',
'registrar_code': '341697',  # course code: 113901
'registrar_code_display_text': 'selenium_test',
```

Values for the test course are:
```
course_instance_id=341697
registrar_code_display=selenium_test
registrar_code=113901
```

I'm not sure what the `109581` corresponds to in the test data above, and the labels attached to the values in the test data don't match what our database record/columns show, so connecting the two either needs copious comments or the variables themselves could be renamed (preferable, IMO).

As for the tests themselves, it was my undestanding that we just need two tests, both of which follow this pattern (for the tests, assume course instance A has a registrar_code_display, and course instance B does not):
1. search for the course instance
2. check the value in the 'course code' column.
  -test A: value should match registrar_code_display
  -test B: value should match registrar_code

There's no need for a non-zero field length check -- the tests in no.2 above automatically cover that case as well.

In that case the test data could really just be split like so:

```
'bulk_create': {
    'test_data': {
        # test course for TLT-2522
        'ci_without_registrar_code_display': {
            'course_instance_id': '...',
            'registrar_code': '...',
        },
        # default course data for bulk create tests
        'course_code': 'Selenium-Automated',  # defaults to ILE
        'course_group': 'Anthropology',
        'course_instance_id': '341697',
        'course_short_title': 'Selenium Auto Test 101',
        'course_title': 'Selenium Automated Test Course 101',
        'registrar_code_display': 'selenium_test',
        'template': 'None',  # No Template
        'term': '1',
    },
```

(Even better would be to move the fields in the root level of `'test data'` into their own subdict, but we'd have to refactor the other tests a little bit for that to work.)

The test case runs similar to what's in here, but we should really only need two test cases (one, if you use a spreadsheet to bring in two rows of data), and it will check that the 'course code' displays the expected text via a single new service method on the PO (that service method simply returns the text in the course code field, or if that's too difficult, it can search in the table for the expected value passed in as an argument).

I think this same logic applies for TLT-2521 as well. I hope that helps.

hktest commented:
On the 3 key comments above: 
1.  Nomenclature - Fixed and renamed a bunch of them, so hopefully that should resolve the confusion.  Part of the reason is that the column itself is called "course_code" whereas in the db, we're referring to the registrar_code, so I made the variables more consistent now ("registrar_code" and "registrar_code_display") since "course_code" means something else in the db too.   

2.  Fixed up the tests.  There are some concerns around the check for the null value, so maybe Josie, you and I could touch base about it.  I cleaned it up so there are 2 clear tests, but also added the null check as part of the asserts in the test.   Maybe that makes a little more sense this way?  

3.  Cleaned up local.py.  After clean-up, there are really only two new variables added as part of the tests.  Not sure if there is significant value in moving it out into its own subdict at this point vs it staying at the root level - so I haven't moved it out.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -7,25 +7,39 @@
-        self.assertTrue(index_page.is_loaded())
 
+        # Get base_url and switch frame whenever loading up bulk create tool
+        self.driver.get(self.base_url)
```
I think this already happens in the `setUpClass` method of the `BulkCreateBaseTestCase`, so no need for it here -- it will slow down the tests if you have to fetch it again.

hktest replied:
see longer note below.  setupclass does not currently exist for icommons_lti_tools, so this handles it for the tests.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -7,25 +7,39 @@
 
+        # Get base_url and switch frame whenever loading up bulk create tool
+        self.driver.get(self.base_url)
+        index_page.focus_on_tool_frame()
```
You may have to do this here, sure. Or you could try to move it into the `BulkCreateBaseTestCase.setUpClass()` (it would run after the login logic).

hktest replied:
ditto above.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -7,25 +7,39 @@
+        self.driver.get(self.base_url)
+        index_page.focus_on_tool_frame()
+
+        index_page.is_loaded()
```
Did you mean to have a `self.assertTrue()` here? Otherwise this does nothing and can be removed.

hktest replied:
Yes, indeed.  Fixed.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -35,4 +49,59 @@ def test_course_selection(self):
+                registrar_code)
+        expected_registrar_code = registrar_code
+
+        # # Verify that the course_code_field in bulk create is not null
```
As discussed, this is a redundant check.

hktest replied:
Discussed and removed.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -35,4 +49,59 @@ def test_course_selection(self):
+                registrar_code)
+        expected_registrar_code = registrar_code
+
+        # Verify that the course_code_field in bulk create is not null
```
ditto

hktest replied:
Fixed.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -35,4 +49,59 @@ def test_course_selection(self):
+
+    def test_course_code_value_should_match_registrar_code(self):
+        """
+        Tests that a course code field on bulk create tool matches course's "registrar_code"
```
TLT-???? would be helpful in the comments

hktest replied:
Done.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -35,4 +49,59 @@ def test_course_selection(self):
+
+    def test_course_code_value_should_match_registrar_code_display(self):
+        """
+        Tests that a course code field on bulk create tool matches course's
```
TLT-???? would be helpful in the comments

hktest replied:
Done.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -35,4 +49,59 @@ def test_course_selection(self):
+        Tests that a course code field on bulk create tool matches course's
+        "registrar_code_display" value.
+        """
+        self.load_bulk_create_tool() #
```
extra `#` unnecessary here

hktest replied:
Eek. Fixed.

elliottyates commented on selenium_tests/bulk_create/page_objects/index_page.py:
```diff
@@ -16,11 +16,19 @@ class Locators(object):
     CREATE_NEW_COURSE_LINK = (By.ID, "createNewCourseButton")
 
+    @classmethod
+    def COURSE_CODE_TEXT(cls, course_code_text):
```
This doesn't match the course code column -- this finds text in any cell in the table. It can be generalized to `TD_TEXT(cls, text_to_match_in_td)` or something like that.

hktest replied:
Fixed.

elliottyates commented on selenium_tests/bulk_create/page_objects/index_page.py:
```diff
@@ -55,3 +63,17 @@ def get_new_course_link(self):
         element = self.find_element(*Locators.CREATE_NEW_COURSE_LINK)
         element.click()
+
+    def get_content_of_course_code_element(self, registrar_code):
```
Ditto, this is not match course code element, just any cell that has the text. That is a useful service for the PO to provide to the tests, but we should rename the method, change the docstring/comments, and rename the locator. Also note that this is asking to find some text in the table, then returning the same text. Seems like if you're searching by text, you would always expect the same text to come back.

elliottyates commented on selenium_tests/bulk_create/page_objects/index_page.py:
```diff
@@ -55,3 +63,17 @@ def get_new_course_link(self):
+        element = self.find_element(
+            *Locators.COURSE_CODE_TEXT(registrar_code))
+
+        if element.text > 0:
```
I think maybe this should be `if len(element.text) > 0`. And do we want to `strip()` the excess spaces first?

hktest replied:
Fixed!  Yes, we do want to check the length. Modified it to: 
element = element.text.strip()
        if len(element) > 0:
            return element

elliottyates commented:
Also, did we mean to remove roles_access.xlsx? I noticed that disappeared, but not sure why.

elliottyates commented:
:+1:

elliottyates commented:
Also, in general, this could really be one test with a spreadsheet that had different expectations/test data in it.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -7,25 +7,39 @@
+        selection page, and select courses and a template to enable course
+        creation.
+        """
+        course_selection_page = CourseSelectionPageObject(self.driver)
```
Note that if you didn't want to re-instantiate this you could just store them all in `self` in the `load_bulk_create_tool()` or even in the `setUpClass()` in the `BulkCreateBaseTestCase`.

hktest replied:
Ah, took me a while to figure out what this actually means, but fixed (I thought you always have to instantiate in the test method/ learned something new).  I put it in the load_bulk_create_tool() for now - thanks!

hktest commented:
re: missing "roles_access.xlsx"  -- wow, nice catch.  re-added.  I must have fat-fingered deleted.  thanks!

hktest commented:
There are still two pieces of feedback left on this PR - 
1.  We have a setupclass(cls), but not a setupclass (which we don't have for icommons_lti_tools).  It probably would be tidier to add in a new setupclass class in the long run, but the load_bulk_create_tool had been refactored from test_course_selection (unrelated to current story) and the tests are using this new method (and doesn't add too much bulk, is relatively clear,and works too), so I'm going to keep the existing code re: load_bulk_create() to not add scope creep for now.  2. In the meantime, I'm going to focus on cleaning the actual test/comment for this story, per the PR feedback.

hktest commented:
Hi @elliottyates -  thanks for the review.  I implemented as many changes as it made sense per PR feedback.  Since the original tests weren't very solid as it was, I tightened the tests so it's testing precisely what I need it to check and sprinkled comments where it's needed/helpful.  These changes have been tested locally and working, so I plan to merge it in tomorrow unless that are anything particularly concerning otherwise -- thanks!

hktest commented:
Merging - tested bulk_create_tests and also ran the regression_suite.py tests successfully.

elliottyates commented on icommons_lti_tools/settings/local.py:
```diff
@@ -16,6 +16,18 @@
+                'sis_id_value': '341697',
+            },
+            'course_without_registrar_code_display': {
+                'registrar_code_display': '109581',
```
This is a little confusing still -- should this be `'registrar_code'` instead of `'registrar_code_display'`?

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -14,25 +32,73 @@ def test_course_selection(self):
+
+        """
+        self._load_bulk_create_tool()
+        expected_registrar_code = self.test_data_course1[
```
Maybe this should be `expected_course_code`?

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -14,25 +32,73 @@ def test_course_selection(self):
+            'registrar_code_display']
+
+        # This returns the text of a specific locator on bulk_create table
+        registrar_code_element = self.driver.find_element(
```
In the future let's try to keep page elements and driver calls in the page objects. That's the benefit of the page object pattern -- POs encapsulate the details of the page (e.g. the row and column you'd want to check for a specific thing), and expose services like `get_course_code()`. The tests become easier to read and independent of how you code the page.

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -14,25 +32,73 @@ def test_course_selection(self):
+        # This returns the text of a specific locator on bulk_create table
+        registrar_code_element = self.driver.find_element(
+                *Locators.COURSE_CODE_LOCATOR_R4_C3)
+        actual_registrar_code = registrar_code_element.text
```
Ditto registrar_code vs registrar_code_display

elliottyates commented on selenium_tests/bulk_create/bulk_create_tests.py:
```diff
@@ -14,25 +32,73 @@ def test_course_selection(self):
+        """
+        self._load_bulk_create_tool()
+        expected_registrar_code = self.test_data_course2[
+            'registrar_code_display']
```
Ditto `'registrar_code'` vs `'registrar_code_display'`

elliottyates commented on selenium_tests/bulk_create/page_objects/index_page.py:
```diff
@@ -15,12 +15,19 @@ class Locators(object):
     TERM_SELECT_DROPDOWN = (By.ID, "termSelect")
     UNAUTHORIZED_MESSAGE = (By.ID, "unauthorized_message")
     CREATE_NEW_COURSE_LINK = (By.ID, "createNewCourseButton")
+    # This is locating a specific element on bulk create table (row 3/column 3)
```
Additional comments about what we actually semantically expect to be in that space would be helpful.
